clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -98,12 +107,14 @@ int main() {
   // Within a constexpr if, pass the first lambda to a kernel in the true branch, pass the second lambda to a kernel in the false branch
   auto l9 = []() { return 1; };
   auto l10 = []() { return 2; };
+  // expected-note@+1{{'__builtin_unique_stable_name' evaluated here}}
   constexpr const char *l10_output = __builtin_unique_stable_name(l10);
-  if constexpr (1)
+  if constexpr (1) {
     kernel_single_task<class kernel8>(l9);
-  else
+  } else {
+    // expected-note@+1{{in instantiation of function template specialization}}
```

Comment:
There is a new weirdness here regarding errors- 
Notice how instead of adding a check for error such as `expected-error@+1{{kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}`, I am having to use`// expected-note@+1{{in instantiation of function template specialization}}` ? All the `kernel instantiation changes the result of an evaluated __builtin_unique_stable_name` are now occuring at line 5, as expressed here: // expected-error 8 @+2{{kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}} 

I believe that this is because of the instantiating kernel_single_task with different KernelName for my test cases. This turns the error messaging to be unhelpful, or even useless, because this instantiation is the mechanism used even for tests where we don't expect errors. I am working on ways to avoid this, but may be I am missing something else here :( Please share your thoughts! Thank you.

Reply:
You want to keep the instantiation messages together with what caused them if at all possible, and you'd want to do the same with the errors as well (basically group them around where it makes most sense).  

See this page here: https://clang.llvm.org/doxygen/classclang_1_1VerifyDiagnosticConsumer.html
The bit in particular that starts with this: As an alternative to specifying a fixed line number,

Reply:
Thank you. I grouped cases that cause error to one `#ifdef TRIGGER_ERROR` section, and those that don't in the `#else` section. However, this didn't solve the problem. My main issue is that `expected-error{{kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}` shows up at the definition of `kernel_single_task` rather than at its call site, irrespective of whether the specific test case actually causes issues. Every call site is treated as something that is reevaluating the builtin. 
I went through that link but I am not sure how an alternative method of specifying line number can help here.

Reply:
FWIW, Sema tests should rarely need to use `TRIGGER_ERROR` (we do that for some of our tests because we mix ast dumping tests with other semantic tests -- those tests should be split into two, an AST dumping test that lives in AST and the semantic tests that live in SemaSYCL).

The current diagnostic behavior seems valid to me in that the call to `__builtin_unique_stable_name` in the tests is what "fixes" the stable name, and then when the kernel uses the name, that's the point at which we can say the programmer has violated the contract.

> ...irrespective of whether the specific test case actually causes issues.

Can you explain a bit further why you think this test case shouldn't cause diagnostics (or why the diagnostics are in the wrong location)?

Reply:
My main I am concerned about is that I should see no errors for the test cases which use the builtin correctly.  Please see line 32 of SemaSYCL/unique_stable_name.cpp test
`


<!--StartFragment-->// This section is not supposed to emit any diagnostics but there are currently twotemplate &lt;typename KernelName, typename KernelType&gt;[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) { // expected-error 2 {{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}  kernelFunc();}<!--EndFragment-->

`

Reply:
@AaronBallman are you suggesting the removal of the TRIGGER_ERROR altogether?

Reply:
> @AaronBallman are you suggesting the removal of the TRIGGER_ERROR altogether?

Yes, there's no need for it in the test file as it's currently used.